src/devices/machine/qvt102.cpp

```diff
@@ -0,0 +1,692 @@
+	}
+
+	return state;
+}
+
+WRITE8_MEMBER(qvt102_device::mcu_p1_w)
+{
+	m_kbd_p1 = data;
+}
+
+WRITE8_MEMBER(qvt102_device::mcu_p2_w)
+{
+	m_kbd_p2 = data;
+
+	m_kbd_data = !BIT(data, 7);
+	m_irqs->in_w<2>(m_kbd_data);
+}
+
+
+//**************************************************************************
+//  VIDEO EMULATION
+//**************************************************************************
+
+READ8_MEMBER(qvt102_device::vsync_ack_r)
+{
+	m_irqs->in_w<0>(CLEAR_LINE);
+	return 0;
+}
+
+WRITE_LINE_MEMBER(qvt102_device::vsync_w)
+{
+	if (state)
+		m_irqs->in_w<0>(ASSERT_LINE);
+}
+
+MC6845_UPDATE_ROW( qvt102_device::crtc_update_row )
+{
+	const rgb_t *palette = m_palette->palette()->entry_list_raw();
+
+	// line attribute (active for the rest of the line)
+	uint8_t attr = 0;
+
+	for (int x = 0; x < x_count; x++)
+	{
+		uint16_t mem = (ma + x) & 0x7ff;
+		uint8_t chr = m_vram[mem];
+
+		// address lines 10 and 11 for the character rom get a special handling
+		int a10 = (chr & 0x40) || ((chr & 0xe0) == 0);
+		int a11 = 0;
+
+		switch (m_jumper->read() & 0x30)
+		{
+		case 0x00: a11 = 0; break;    // US
+		case 0x10: a11 = a10; break;  // German
+		case 0x20: a11 = !a10; break; // UK
+		}
+
+		uint16_t gfx = m_char_rom[(a11 << 11) | (a10 << 10) | ((chr << 4) & 0x3f0) | ra] << 1;
+
+		// d0 for drawing codes (avoids gap)
+		if (chr >= 0x80 && chr <= 0x9f)
+			gfx |= BIT(gfx, 1);
+
+		// half intensity is defined per character
+		int half = BIT(chr, 7);
+
+		// check for new attribute
+		if (chr >= 0x90 && chr <= 0x9f)
+		{
+			attr = chr;
+			half ^= BIT(m_jumper->read(), 6);
+		}
+
+		// draw 9 pixels of the character
+		for (int i = 0; i < 9; i++)
+		{
+			// from schematics
+			bool p = !(BIT(m_latch, 3) & BIT(attr, 1)); // blink
+			p = !(p & BIT(gfx, i) & (!BIT(attr, 0))); // rom/blank
+			p = !(p & !((BIT(attr, 3) & (ra == 11)))); // underline
+			p = p ^ (BIT(attr, 2) ^ ((x == cursor_x) | BIT(m_latch, 2))); // reverse
+
+			bitmap.pix32(y, x*9 + (8-i)) = palette[p ? 2 - half : 0];
+		}
+
+	}
+}
+
+static const gfx_layout char_layout =
+{
+	8,12,
+	RGN_FRAC(1,1),
+	1,
+	{ 0 },
+	{ 0, 1, 2, 3, 4, 5, 6, 7 },
+	{ 0*8, 1*8, 2*8, 3*8, 4*8, 5*8, 6*8, 7*8, 8*8, 9*8, 10*8, 11*8 },
+	8*16
+};
+
+static GFXDECODE_START(chars)
+	GFXDECODE_ENTRY("chargen", 0, char_layout, 0, 1)
+GFXDECODE_END
+
+
+//**************************************************************************
+//  MACHINE EMULATION
+//**************************************************************************
+
+WRITE8_MEMBER(qvt102_device::latch_w)
+{
+	// 7------- host to keyboard
+	// -6------ aux print
+	// --5----- ctc cs0
+	// ---4---- dtri
+	// ----3--- blink
+	// -----2-- rv
+	// ------1- eia/cl
+	// -------0 aux en
+
+	m_latch = data;
+
+	// jumper w3 enables this connection
+	if (machine().phase() != machine_phase::INIT)
+	{
+		if (BIT(m_jumper->read(), 2))
+		{
+			fprintf(stderr, "######## rs232_write_dtr %02x\n", BIT(data, 4));
```

Comment:
Please use `logerror` for general diagnostic messages, and don't put the hashes.  The device tag will autmatically be logged that way, it will be visible to the debugger, and you can send it to the console with `-oslog`.

Why are you logging a single bit as a two-digit hex number?

Also, this should probably use `LOG` from `logmacro.h` to put it behind the `VERBOSE` macro as it isn't really something unusual.